File: Discrete_Fourier_Transform/discrete_fourier_transform.m
clc;clear all; close all; 
 
%Code by: circuitpotato
%Visit downtothecircuits.wordpress.com for more information
 
%Purpose: This code example shows the conversion of a time domain sampled 
%signal to frequency domain using Discrete Fourier Transform
 
%Input Specifications 
N = 1000;   % samples
Fs = 40e3;  % Nyquist Frequency / Sampling Frequency
f1 = 13e3;  % signal frequency 1
f2 = 10e3;  % signal Frequency 2
amp_1 = 3;  % signal amplitude 1
amp_2 = 1;  % signal amplitude 2
 
%White noise specifications
mu = 0;     % mean
std = 1;    % standard deviation
white_noise = std*randn(1,N)+mu;
 
%x-axis Specifications (Best not to change)
Ts = 1/Fs;  % Sampling Period 
dt = 0:Ts:(N-1)*Ts; % start:step:stop
 
x = amp_1*sin(2*pi*f1*dt)+amp_2*sin(2*pi*f2*dt)+white_noise;  % required signal + white noise
 
ReX = zeros(N/2,1);
ImX = zeros(N/2,1);
for k = 0:(N-1)/2
  for n = 0:N-1
    ReX(k+1) = ReX(k+1) + x(n+1)*cos((2*pi*k*n)/N); % real part
    ImX(k+1) = ImX(k+1) + x(n+1)*sin((2*pi*k*n)/N); % imaginary part 
  end
end
 
magnitude=zeros(N/2,1);
for q=1:N/2
  magnitude(q) = sqrt(ReX(q)^2+ImX(q)^2);
end
 
y_axis = magnitude./(N/2);   % peak amplitude values
x_axis = 0:Fs/N:((N/2)-1)*(Fs/N);   % frequency scale 
 
figure(1);
plot(x);
title('time domain sampled signal');
xlabel('Samples');
ylabel('Amplitude');
 
figure(2);
plot(x_axis,y_axis)
xlabel('Frequency');
ylabel('Amplitude');
title('Frequency Spectrum');

File: Discrete_Fourier_Transform/discrete_fourier_transform.py
import numpy as np
import matplotlib.pyplot as plt
 
#Code by: circuitpotato
#Visit downtothecircuits.com for more information
 
#Purpose: This code example shows the conversion of a time domain sampled
#signal to frequency domain using Discrete Fourier Transform
 
#Input Specifications
N = 1000;   # samples
Fs = 40e3;  # Nyquist Frequency / Sampling Frequency
f1 = 13e3;  # signal frequency 1
f2 = 10e3   # signal frequency 2
amp_1 = 3;  # signal amplitude 1
amp_2 = 1   # signal amplitude 2
 
#White Noise Specification
mu = 0  # mean
std = 1 # standard deviation
white_noise = np.random.normal(mu, std, size=N)
 
#x-axis Specifications (Best not to change)
Ts=1/Fs   # Sampling period
dt=np.arange(0,(N)*Ts,Ts)   # np.arange(start,stop,step)
 
x=amp_1*np.sin(2*np.pi*f1*dt)+amp_2*np.sin(2*np.pi*f2*dt)+white_noise   # required signal + white noise
 
ReX = np.zeros((int(N/2),1))
ImX = np.zeros((int(N/2),1))
for k in range(0,int(N/2)):
    for n in range(0,N):
        ReX[k] = ReX[k] + x[n]*np.cos((2*np.pi*k*n)/N)  # real part
        ImX[k] = ImX[k] + x[n]*np.sin((2*np.pi*k*n)/N)  # imaginary part
 
magnitude=np.zeros((int(N/2),1))
for q in range(0,int(N/2)):
    magnitude[q]=np.sqrt((ReX[q]**2)+(ImX[q]**2))
 
y_axis=magnitude/(N/2)   # peak amplitude values
x_axis=np.arange(0,(N/2)*(Fs/N),(Fs/N)) # frequency values
 
plt.figure(1)
plt.plot(x)
plt.title('time domain sampled signal')
plt.xlabel('Samples')
plt.ylabel('Amplitude')
 
plt.figure(2)
plt.plot(x_axis,y_axis)
plt.xlabel('Frequency')
plt.ylabel('Amplitude')
plt.title('Frequency Spectrum')
 
plt.show()
